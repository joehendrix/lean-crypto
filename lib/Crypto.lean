import Crypto.Mceliece.Ref348864
import Crypto.Prim.AES
import Crypto.Prim.Shake
import Crypto.Random.DRBG