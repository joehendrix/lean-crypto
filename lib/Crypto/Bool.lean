namespace Bool

instance : Xor Bool := ⟨bne⟩

@[simp] theorem false_xor (a : Bool) : false ^^^ a = a := by cases a <;> rfl
@[simp] theorem xor_false (a : Bool) : a ^^^ false = a := by cases a <;> rfl
@[simp] theorem true_xor (a : Bool) : true ^^^ a = !a := by cases a <;> rfl
@[simp] theorem xor_true (a : Bool) : a ^^^ true = !a := by cases a <;> rfl

theorem xor_comm (a b : Bool) : a ^^^ b = b ^^^ a :=
  by cases a <;> simp

theorem xor_assoc (a b c : Bool) : a ^^^ b ^^^ c = a ^^^ (b ^^^ c) :=
  by cases a <;> cases b <;> simp

-- TODO(WN): Should these be in the Lean namespace? And probably Mathlib classes
-- should imply them.
instance : Lean.IsCommutative (α := Bool) HXor.hXor := ⟨xor_comm⟩

instance : Lean.IsAssociative (α := Bool) HXor.hXor := ⟨xor_assoc⟩

end Bool