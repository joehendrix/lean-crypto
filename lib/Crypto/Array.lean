import Crypto.Fin

theorem ite_same {α:Type} (p:Prop) [h:Decidable p] (a:α) : (if p then a else a) = a := by
  cases h with
  | isFalse q =>
     simp only [q]
     trivial
  | isTrue q =>
    simp only [q]
    trivial

namespace Array

theorem size_mkEmpty {α: Type _} (n:Nat) : size (mkEmpty n : Array α) = 0 := rfl

def generate (n:Nat) (f : Fin n → α) : Array α :=
  foreachFin n (λi a => a.push (f i)) (mkEmpty n)

theorem size_generate {α} (n:Nat) (f:Fin n → α) : (generate n f).size = n := by
  simp only [generate, foreachFin]
  have h : ∀(i:Nat) (p:i ≤ n) (a:Array α),
             size (foreachFin.loop n (fun i a => push a (f i)) i p a) = a.size + i
              := by
    intro i p
    induction i with
    | zero =>
      intro a
      simp only [foreachFin.loop]
      trivial
    | succ i h =>
      intro a
      simp only [foreachFin.loop, h, size_push]
      simp only [Nat.succ_add]
      trivial
  apply Eq.trans (h _ _ _)
  simp only [size_mkEmpty, Nat.zero_add]

theorem size_append {α:Type} : ∀(x y:Array α), (x ++ y).size = x.size + y.size
| a, b => by
  simp only [HAppend.hAppend, Append.append, Array.append]
  simp only [foldl, foldlM]
  simp
  have h : ∀{j k:Nat} (_ : j+k = b.size) (r:Array α) (p: b.size ≤ b.size),
            size (Id.run (foldlM.loop (fun r v => push r v) b b.size p j k r))
              = r.size + j := by
    intro j k q r p
    revert k q r
    induction j with
    | zero =>
      intro k q r
      unfold foldlM.loop
      have not_q : ¬(k < b.size) := by
        rw [q.symm]
        simp [ Nat.lt_irrefl]
      simp [not_q]
      trivial
    | succ j ind =>
      intro k q r
      unfold foldlM.loop
      have is_lt : k < b.size := by
        rw [Nat.add_comm] at q
        exact Nat.eq_add_implies_lt q
      simp [is_lt]
      apply Eq.trans
      have q_ind : j + (k + 1) = size b := by
        simp only [Nat.succ_add] at q
        exact q
      apply ind q_ind
      simp only [size_push]
      simp only [Nat.succ_add]
      trivial
  apply h
  trivial

@[simp]
theorem size_ofSubarray {α:Type} (s:Subarray α) : (ofSubarray s).size = s.stop - s.start := sorry

@[simp]
theorem start_toSubarray {α:Type} (a:Array α) (start stop : Nat) (p : start ≤ stop) (q : stop ≤ a.size)
  : (toSubarray a start stop).start = start := by simp [toSubarray, p, q]

@[simp]
theorem start_toSubarray_large {α:Type} (a:Array α) (start stop : Nat) (p : a.size ≤ start) (q : a.size ≤ stop)
  : (toSubarray a start stop).start = a.size := by
    simp only [toSubarray]
    cases Decidable.em (stop ≤ size a) with
    | inl g =>
      simp [g]
      cases Decidable.em (start ≤ stop) with
      | inl h =>
        simp [h]
        have r : start ≤ size a := Nat.le_trans h g
        exact (Nat.le_antisymm r p)
      | inr h =>
        simp [h]
        exact (Nat.le_antisymm g q)
    | inr g =>
      simp [g]
      cases Decidable.em (start ≤ a.size) with
      | inl h =>
        simp [h]
        exact (Nat.le_antisymm h p)
      | inr h =>
        simp [h]

theorem stop_toSubarray_large {α:Type} (a:Array α) (start stop : Nat) (_ : a.size ≤ start) (q : a.size ≤ stop)
  : (toSubarray a start stop).stop = a.size := by
    simp only [toSubarray]
    cases Decidable.em (stop ≤ size a) with
    | inl g =>
      simp [g]
      cases Decidable.em (start ≤ stop) with
      | inl h =>
        simp [h]
        exact (Nat.le_antisymm g q)
      | inr h =>
        simp [h]
        exact (Nat.le_antisymm g q)
    | inr g =>
      simp [g]
      cases Decidable.em (start ≤ a.size) with
      | inl h => simp [h]
      | inr h => simp [h]

@[simp]
theorem stop_toSubarray {α:Type} (a:Array α) (start stop : Nat) (p : start ≤ stop) (q : stop ≤ a.size)
  : (toSubarray a start stop).stop = stop := by simp [toSubarray, p, q]

theorem size_empty {α:Type _} : Array.size (#[] : Array α) = 0 := by rfl

@[simp]
theorem size_extract {α:Type} (src: Array α) (start stop : Nat)
: (src.extract start stop).size = min stop src.size - start := by
  simp [extract, toSubarray]
  cases Decidable.em (stop ≤ size src) with
  | inl g =>
    simp [g, min]
    cases Decidable.em (start ≤ stop) with
    | inl h => simp [h]
    | inr h =>
      have q := Nat.gt_of_not_le h
      simp [h, Nat.sub_self, Nat.lt_implies_zero_sub q]
  | inr g =>
    simp [g, min]
    cases Decidable.em (start ≤ size src) with
    | inl h =>
      simp [h]
    | inr h =>
      have q := Nat.gt_of_not_le h
      simp [h, Nat.sub_self, Nat.lt_implies_zero_sub q]

theorem extract_all (a:Array α)
  : (a.extract 0 a.size) = a := by
    admit

theorem extract_end_empty {a:Array α} {i : Nat} (p : i ≥ a.size) (j : Nat)
  :(a.extract i j) = Array.empty := by
    admit


theorem append_empty {α} (a:Array α) : a ++ empty = a := by
  admit

theorem append_data {α} (a b:Array α) : (a ++ b).data = a.data ++ b.data := by
  admit

theorem size_qsort {α} [Inhabited α] (a:Array α) (lt : α → α → Bool)
  : Array.size (Array.qsort a lt) = a.size := by
  admit

@[simp]
theorem size_setD (a:Array α) : (a.setD i e).size = a.size := by
  simp only [setD]
  cases Decidable.em (i < a.size) with
    | inl h => simp [h]
    | inr h => simp [h]

theorem size_set! (a:Array α) : (a.set! i e).size = a.size := by
  simp [set!, size_setD]

end Array
