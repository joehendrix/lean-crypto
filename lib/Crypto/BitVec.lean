import Crypto.Range
import Crypto.UInt8
import Crypto.ToMathlib

import Smt.Data.BitVec

namespace BitVec

def get_lsb (x : BitVec m) (i : Fin m) : Bool :=
  (x.val &&& (1 <<< i.val)) ≠ 0

def lsb_getAux (x : BitVec m) (i : Nat) : Bool :=
  (x.val &&& (1 <<< i)) ≠ 0

@[implementedBy lsb_getAux]
def lsb_get! (x : BitVec m) (i : Nat) : Bool :=
  BitVec.lsbGet x i

def lsb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  if c then
    x ||| ⟨1 <<< i, sorry⟩
  else
    x &&& ⟨((1 <<< m) - 1 - (1 <<< i)), sorry⟩

def get_msb (x : BitVec m) (i : Fin m) : Bool :=
  (x.val &&& (1 <<< (m-1-i.val))) ≠ 0

def msb_get! (x : BitVec m) (i : Nat) : Bool :=
  (x.val &&& (1 <<< (m-1-i))) ≠ 0

def msb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  let i  := m-1-i
  if c then
    x ||| ⟨1 <<< i, sorry⟩
  else
    x &&& ⟨((1 <<< m) - 1 - (1 <<< i)), sorry⟩

/--
Update index to use most-significant bytes, but least-significant bit
ordering within bytes.

This may be removed once compatibility with C is not needed.
-/
def msbb_fix (m : Nat) (i : Nat) : Nat :=
  let j := (m-1)-i
  -- Reverse bit order within bytes (see if we can fix this)
  ((j >>> 3) <<< 3) ||| (0x7 - (j &&& 0x7))

def msbb_get! (x : BitVec m) (i : Nat) : Bool := x.lsb_get! (msbb_fix m i)

protected def toBinary (x : BitVec n) : String :=
  let l := Nat.toDigits 2 x.val
  "0b" ++ String.mk (List.replicate (n - l.length) '0' ++ l)

protected def toHex (x : BitVec n) : String :=
  let l := Nat.toDigits 16 x.val
  "0x" ++ String.mk (List.replicate (n/4 - l.length) '0' ++ l)

def reverse (x : BitVec n) : BitVec n := Id.run do
  let mut r : Nat := 0
  for i in range 0 n do
    r := r <<< 1
    if x.lsb_get! i then
      r := r + 1
  pure ⟨r, sorry⟩

-- Fold that visits least-significant bit first.
protected def fold_lsb (f : α → Bool → α) (x : BitVec n) (a : α) : α := Id.run do
  let mut r := a
  let mut v := x.val
  for _i in range 0 n do
    r := f r ((v &&& 1) = 1)
    v := v >>> 1
  pure r

protected def take_lsb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val &&& 1 <<< n - 1, sorry⟩

protected def take_msb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val >>> (m-n), sorry⟩

def extractN! (a:BitVec n) (s m:Nat) : BitVec m :=
  let e := s + m
  let b := (a.val >>> (n - e)) &&& (1 <<< (min m (n - s)) - 1)
  ⟨b, sorry⟩

theorem eq_of_val_eq {n:Nat} : ∀{x y : BitVec n}, x.val = y.val → x = y
  | ⟨_,_⟩, _, rfl => rfl

theorem ne_of_val_ne {x y : BitVec n} (h : x.val ≠ y.val) : x ≠ y :=
  λ h' => absurd (h' ▸ rfl) h

protected def decideEq : (x y : BitVec n) → Decidable (x = y) :=
  fun ⟨i, _⟩ ⟨j, _⟩ =>
    match decEq i j with
    | isTrue h  => isTrue (eq_of_val_eq h)
    | isFalse h => isFalse (ne_of_val_ne h)

instance : DecidableEq (BitVec n) := BitVec.decideEq

protected def generate_lsb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  for i in range 0 n do
    let b := f ⟨i, sorry⟩
    r := (if b then 1 <<< i else 0) ||| r

  ⟨r, sorry⟩

protected def generate_msb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  for i in range 0 n do
    let b := f ⟨i, sorry⟩
    r := r <<< 1 ||| (if b then 1 else 0)

  ⟨r, sorry⟩

  def toUInt8  (x:BitVec  8) : UInt8  := OfNat.ofNat x.val
  def toUInt16 (x:BitVec 16) : UInt16 := OfNat.ofNat x.val
  def toUInt32 (x:BitVec 32) : UInt32 := OfNat.ofNat x.val
  def toUInt64 (x:BitVec 64) : UInt64 := OfNat.ofNat x.val

  protected def toString {n:Nat} (x:BitVec n) : String :=
    if n % 16 = 0 then
      let s := (Nat.toDigits 16 x.val).asString
      let t := (List.repeat' '0' (n / 16 - s.length)).asString
      "0x" ++ t ++ s
    else
      let s := (Nat.toDigits 2 x.val).asString
      let t := (List.repeat' '0' (n - s.length)).asString
      "0b" ++ t ++ s

  instance : ToString (BitVec n) := ⟨BitVec.toString⟩

end BitVec

namespace UInt8
  def toBitVec (x:UInt8) : BitVec 8 := ⟨x.toNat, x.val.isLt⟩
end UInt8

namespace UInt16
  def toBitVec (x:UInt16) : BitVec 16 := ⟨x.toNat, x.val.isLt⟩
end UInt16

namespace UInt32
  def toBitVec (x:UInt32) : BitVec 32 := ⟨x.toNat, x.val.isLt⟩
end UInt32

namespace UInt64
  def toBitVec (x:UInt64) : BitVec 64 := ⟨x.toNat, x.val.isLt⟩
end UInt64