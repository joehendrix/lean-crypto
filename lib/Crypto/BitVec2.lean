import Crypto.Range
import Crypto.UInt8
import Crypto.ToMathlib

import Smt.Data.BitVec

namespace BitVec

def lsb_getAux (x : BitVec m) (i : Nat) : Bool :=
  (x.val &&& (1 <<< i)) ≠ 0

@[implementedBy lsb_getAux]
def lsb_get! (x : BitVec m) (i : Nat) : Bool :=
  BitVec.lsbGet x i

def lsb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  if c then
    x ||| ⟨1 <<< i, sorry⟩
  else
    x &&& ⟨((1 <<< m) - 1 - (1 <<< i)), sorry⟩

/--
Update index to use most-significant bytes, but least-significant bit
ordering within bytes.

This may be removed once compatibility with C is not needed.
-/
def msbb_fix (m : Nat) (i : Nat) : Nat :=
  let j := (m-1)-i
  -- Reverse bit order within bytes (see if we can fix this)
  ((j >>> 3) <<< 3) ||| (0x7 - (j &&& 0x7))

def msbb_get! (x : BitVec m) (i : Nat) : Bool := x.lsb_get! (msbb_fix m i)

def msbb_set! (x : BitVec m) (i : Nat) (c : Bool) : BitVec m :=
  x.lsb_set! (msbb_fix m i) c

protected def toBinary (x : BitVec n) : String :=
  let l := Nat.toDigits 2 x.val
  "0b" ++ String.mk (List.replicate (n - l.length) '0' ++ l)

protected def toHex (x : BitVec n) : String :=
  let l := Nat.toDigits 16 x.val
  "0x" ++ String.mk (List.replicate (n/4 - l.length) '0' ++ l)

protected def toHex2 (x : BitVec n) : String := Id.run do
  let mut s : String := "0x"
  for i in range 0 (n/8) do
    let b := UInt8.ofNat (x.val >>> (8*i))
    s := s ++ b.toHex
  pure s

instance : ToString (BitVec n) := ⟨BitVec.toHex2⟩

def reverse (x : BitVec n) : BitVec n := Id.run do
  let mut r : Nat := 0
  for i in range 0 n do
    r := r <<< 1
    if x.lsb_get! i then
      r := r + 1
  pure ⟨r, sorry⟩

protected def foldl (f : α → Bool → α) (x : BitVec n) (a : α) : α := Id.run do
  let mut r := a
  for i in range 0 n do
    r := f r (x.msbb_get! i)
  pure r

protected def take_lsb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val &&& 1 <<< n - 1, sorry⟩

protected def take_msb (x : BitVec m) (n : Nat) : BitVec n :=
  ⟨x.val >>> (m-n), sorry⟩

theorem eq_of_val_eq : ∀ {x y : BitVec n}, x.val = y.val → x = y
  | ⟨_,_⟩, _, rfl => rfl

theorem ne_of_val_ne {x y : BitVec n} (h : x.val ≠ y.val) : x ≠ y :=
  λ h' => absurd (h' ▸ rfl) h

protected def decideEq : (x y : BitVec n) → Decidable (x = y) :=
  fun ⟨i, _⟩ ⟨j, _⟩ =>
    match decEq i j with
    | isTrue h  => isTrue (eq_of_val_eq h)
    | isFalse h => isFalse (ne_of_val_ne h)

instance : DecidableEq (BitVec n) := BitVec.decideEq

protected def generate_lsb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  for i in range 0 n do
    let b := f ⟨i, sorry⟩
    r := (if b then 1 <<< i else 0) ||| r

  ⟨r, sorry⟩

protected def generate_msbb (n : Nat) (f : Fin n → Bool) : BitVec n := Id.run do
  let mut r : Nat := 0

  let m := n % 8
  if m ≠ 0 then
    let mut w : Nat := 0
    for j in range 0 m do
      let b := f ⟨j, sorry⟩
      w := if b then 1 <<< j ||| w else w
    r := w

  for i in range 0 (n/8) do
    let mut w : Nat := 0
    for j in range 0 8 do
      let b := f ⟨m+8*i+j, sorry⟩
      w := if b then 1 <<< j ||| w else w
    r := r <<< 8 ||| w

  ⟨r, sorry⟩

end BitVec
