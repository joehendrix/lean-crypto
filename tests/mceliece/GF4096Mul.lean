import Crypto
import Smt

/-! Optimized Galois field operations in McEliece are the same as generic reference algorithm. -/

namespace Mceliece.Ref348864

def GF4096 : GaloisField2k where
  k := 12
  -- x^12 + x^3 + 1
  irred := BitVec.ofNat 12 0b1001

def GF4096.ofGF (x : GF) : GF4096 :=
  BitVec.shrink (w := 16) 12 x.val.val

open Smt Term Translator in
/-- Translate `{x : α // _}` as `α`. -/
@[smtTranslator]
def translateSubtype : Translator
  | .app (.app (.const ``Subtype _) e) _ =>
    applyTranslators! e
  | .app (.app (.app (.const ``Subtype.val _) _) _) e => do
    applyTranslators! e
  | .app (.app (.app (.app (.const ``Subtype.mk _) _) _) e) _ => do
    applyTranslators! e
  | _ => return none

open Lean Smt Term Translator in
/-- Translate `UIntXY` as `BitVec XY`.
It may be better to use an isomorphism-transfer tactic first, and then just the standard `BitVec`
translation. -/
@[smtTranslator]
def translateUInt : Translator
  | .const ``UInt16 _ =>
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT "16")
  | .app (.const ``UInt16.val _) e =>
    applyTranslators! e
  | e@(.app (.const ``UInt16.ofNat _) n) => do
    let n ← reduceLit n e
    return Smt.BitVec.mkLit 16 n
  | .const ``UInt16.add _        => return symbolT "bvadd"
  | .const ``UInt16.sub _        => return symbolT "bvsub" -- QF_BV
  | .const ``UInt16.mul _        => return symbolT "bvmul"
  | .const ``UInt16.div _        => return symbolT "bvudiv"
  | .const ``UInt16.mod _        => return symbolT "bvurem"
  | .const ``UInt16.modn _       => throwError "unknown"
  | .const ``UInt16.land _       => return symbolT "bvand"
  | .const ``UInt16.lor _        => return symbolT "bvor"
  | .const ``UInt16.xor _        => return symbolT "bvxor"
  | .const ``UInt16.shiftLeft _  => return symbolT "bvshl"
  | .const ``UInt16.shiftRight _ => return symbolT "bvlshr"
  | .app (.const ``UInt16.toUInt32 _) e =>
    return mkApp2 (symbolT "concat") (BitVec.mkLit 16 0) (← applyTranslators! e)
  | .const ``UInt32 _ =>
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT "32")
  | .app (.const ``UInt32.val _) e =>
    applyTranslators! e
  | e@(.app (.const ``UInt32.ofNat _) n) => do
    let n ← reduceLit n e
    return BitVec.mkLit 32 n
  | .const ``UInt32.add _        => return symbolT "bvadd"
  | .const ``UInt32.sub _        => return symbolT "bvsub" -- QF_BV!
  | .const ``UInt32.mul _        => return symbolT "bvmul"
  | .const ``UInt32.div _        => return symbolT "bvudiv"
  | .const ``UInt32.mod _        => return symbolT "bvurem"
  | .const ``UInt32.modn _       => throwError "unknown"
  | .const ``UInt32.land _       => return symbolT "bvand"
  | .const ``UInt32.lor _        => return symbolT "bvor"
  | .const ``UInt32.xor _        => return symbolT "bvxor"
  | .const ``UInt32.shiftLeft _  => return symbolT "bvshl"
  | .const ``UInt32.shiftRight _ => return symbolT "bvlshr"
  | .app (.const ``UInt32.toUInt16 _) e =>
    return appT
      (mkApp3 (symbolT "_") (symbolT "extract") (literalT "15") (literalT "0"))
      (← applyTranslators! e)
  | _ => return none
where
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

open GF2BVPoly

set_option maxHeartbeats 400000 in
example (x y : GF) : GF4096.ofGF (x * y) = (GF4096.ofGF x) * (GF4096.ofGF y) := by
  show (GF4096.ofGF (GF.mul x y)) = GaloisField2k.mul (GF4096.ofGF x) (GF4096.ofGF y)

  extract_def GF4096.ofGF
    specialize_def GF4096.ofGF [] blocking [@Subtype.val, UInt16.val]
    rw [GF4096.ofGF.specialization]
    rw [GF4096.ofGF.specialization]
    rw [GF4096.ofGF.specialization]
    save

  extract_def GF.mul
    specialize_def Mceliece348864Ref.GF.mul [] blocking [
      @Subtype.val,
      UInt16.toUInt32,
      GF.red
    ]
    save

  extract_def GaloisField2k.mul
    specialize_def GaloisField2k.mul [GF4096] blocking [@polyMod, @polyMul]
      rw [GaloisField2k.mul.GF4096.specialization]
      save

  smt_show [
    GF4096.ofGF,
    GF,
    GF.mul,
    GF.red,
    gfMask
  ]

  sorry

#exit -- times out below

  extract_def polyMod
  specialize_def GF2BVPoly.polyMod [24, 12]
  save

  extract_def polyMul
  specialize_def GF2BVPoly.polyMul [12, 12]
  save
#exit
  conv at GaloisField2k.mul.GF4096.def =>
    intro a b
    rw [ GF2BVPoly.polyMul.«12».«12».specialization,
      GF2BVPoly.polyMod.«24».«12».specialization ]
  save

  smt_show [
    GF4096.ofGF,
    GF,
    GF.mul,
    GF.red,
    gfMask,

    GaloisField2k.mul.GF4096,
    GF2BVPoly.polyMul.«12».«12»,
    GF2BVPoly.polyMod.«24».«12»
  ]

  sorry
