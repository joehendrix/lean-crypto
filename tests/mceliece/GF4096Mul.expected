goal: GF4096.ofGF (GF.mul x y) = GaloisField2k.mul.GF4096 (GF4096.ofGF x) (GF4096.ofGF y)

query:
(define-sort Mceliece348864Ref.GF () (_ BitVec 16))
(define-fun Mceliece348864Ref.gfMask () (_ BitVec 16) #b0000111111111111)
(define-fun Mceliece348864Ref.GF.red ((tmp (_ BitVec 32))) Mceliece348864Ref.GF (let ((t (bvand tmp #b00000000011111111100000000000000))) (let ((tmp (bvxor (bvxor tmp (bvlshr t #b00000000000000000000000000001001)) (bvlshr t #b00000000000000000000000000001100)))) (let ((t (bvand tmp #b00000000000000000011000000000000))) (let ((tmp (bvxor (bvxor tmp (bvlshr t #b00000000000000000000000000001001)) (bvlshr t #b00000000000000000000000000001100)))) (bvand ((_ extract 15 0) tmp) Mceliece348864Ref.gfMask))))))
(define-fun Mceliece348864Ref.GF.mul ((x Mceliece348864Ref.GF) (y Mceliece348864Ref.GF)) Mceliece348864Ref.GF (let ((v (bvxor #b00000000000000000000000000000000 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000000)))))) (let ((v_1 (bvxor v (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000001)))))) (let ((v_2 (bvxor v_1 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000010)))))) (let ((v_3 (bvxor v_2 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000011)))))) (let ((v_4 (bvxor v_3 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000100)))))) (let ((v_5 (bvxor v_4 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000101)))))) (let ((v_6 (bvxor v_5 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000110)))))) (let ((v_7 (bvxor v_6 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000000111)))))) (let ((v_8 (bvxor v_7 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000001000)))))) (let ((v_9 (bvxor v_8 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000001001)))))) (let ((v_10 (bvxor v_9 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000001010)))))) (let ((v_11 (bvxor v_10 (bvmul (concat #b0000000000000000 x) (bvand (concat #b0000000000000000 y) (bvshl #b00000000000000000000000000000001 #b00000000000000000000000000001011)))))) (Mceliece348864Ref.GF.red (ite (>= 0 12) #b00000000000000000000000000000000 (ite (>= (+ 0 1) 12) v (ite (>= (+ (+ 0 1) 1) 12) v_1 (ite (>= (+ (+ (+ 0 1) 1) 1) 12) v_2 (ite (>= (+ (+ (+ (+ 0 1) 1) 1) 1) 12) v_3 (ite (>= (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 12) v_4 (ite (>= (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 12) v_5 (ite (>= (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 12) v_6 (ite (>= (+ (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 1) 12) v_7 (ite (>= (+ (+ (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 1) 1) 12) v_8 (ite (>= (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 12) v_9 (ite (>= (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 12) v_10 (ite (>= (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ 0 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 1) 12) v_11 v_11)))))))))))))))))))))))))))
(declare-const x Mceliece348864Ref.GF)
(define-fun GF4096.ofGF ((x Mceliece348864Ref.GF)) (_ BitVec 12) (ite (<= (+ 0 1) 12) ((_ extract 11 0) x) #b000000000000))
(declare-const y Mceliece348864Ref.GF)
(declare-fun GaloisField2k.mul.GF4096 ((_ BitVec 12) (_ BitVec 12)) (_ BitVec 12))
(assert (not (= (GF4096.ofGF (Mceliece348864Ref.GF.mul x y)) (GaloisField2k.mul.GF4096 (GF4096.ofGF x) (GF4096.ofGF y)))))
(check-sat)
tests/mceliece/GF4096Mul.lean:86:0: warning: declaration uses 'sorry'
